### PR TITLE
Replaced CaptureScreenshot with a DOM Query

### PR DESCRIPTION
**Pull request by daneroo**
This replaces the implementation of navToEnd which was using CaptureScreenshot to detect scrolling to the end of the Main/Album page. It is replaced with a query to the DOM.

When in the Main/Album Page, the DOM contains `<a href=".." />` elements for all visible images.
`lastPhotoInDOM()` simply returns the last such `href` in document order.

The DOM actually contains more images than those that are visible, in a kind of virtual scrolling window.

In the DOM, but not reflecting exactly the visible photos (actually a superset of the visible elements):
```html
<a href="./photo/AF1QipAAAAAA" aria=label="Photo - Portrait - Jul 15, 2010, 2:10:48 PM"/>
<a href="./photo/AF1QipBBBBBB" aria-label="Photo - Landscape - Jul 15, 2010, 2:03:10 PM"/>
<a href="./photo/AF1QipCCCCCC" aria-label="Video - Landscape - Jul 30, 2010, 7:20:22 PM"/>
...
```

We tried to find the actual *oldest* photo by using the aria-label attribute which contains a date for the photo, unfortunately, that label is localised for each user's language which makes the date format very hard to parse.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -280,54 +278,70 @@ func (s *Session) firstNav(ctx context.Context) error {
...
-		if len(ids) > 0 {
-			if *verboseFlag {
-				log.Printf("We are ready, because element %v is selected", ids[0])
-			}
-			break
-		}
-		time.Sleep(tick)
-	}
+	// wait for page to be loaded
+	chromedp.WaitReady("body", chromedp.ByQuery).Do(ctx)
```
Comment (mpl) [inline main.go]: I'm surprised we can get away with simply removing all that, as it seemed to me the pagedown and end event were getting lost into oblivion if I didn't wait for the page to be loaded before sending them.
But I'll try it out and will let you know.

File: main.go [hunk 2]
```diff
@@ -280,54 +278,70 @@ func (s *Session) firstNav(ctx context.Context) error {
...
+		previousHref = lastHrefInDOM
 	}
 
-	if *verboseFlag {
-		log.Printf("Successfully jumped to the end")
+	// Now that we have stopped scrolling, select (focus) on the last element
+	// The element must be focused, so that navToLast can send "\n" to enter photo detail page
+	lastEltSel := fmt.Sprintf(`a[href="%s"]`, previousHref)
+	if err := chromedp.Focus(lastEltSel).Do(ctx); err != nil {
+		log.Printf("Error focus: %s", lastEltSel)
```
Comment (mpl) [inline main.go]: let's return this error instead of just logging it. The caller will log it.

File: main.go [hunk 3]
```diff
@@ -280,54 +278,70 @@ func (s *Session) firstNav(ctx context.Context) error {
...
+//  <a href="./photo/AF1QipAAAAAA" aria=label="Photo - Portrait - Jul 15, 2010, 2:10:48 PM"/>
+//  <a href="./photo/AF1QipBBBBBB" aria-label="Photo - Landscape - Jul 15, 2010, 2:03:10 PM"/>
+//  <a href="./photo/AF1QipCCCCCC" aria-label="Video - Landscape - Jul 30, 2010, 7:20:22 PM"/>
+// We tried to find the actual *oldest* photo by using the aria-label attribute which contains a date for the photo,
+// unfortunately that label is localised for each user's language which makes the date format very hard to parse.
+func lastPhotoInDOM(ctx context.Context) (string, error) {
+	sel := `a[href^="./photo/"]` // css selector for all links to images with href prefix "./photo/..."
+	var attrs []map[string]string
+	if err := chromedp.AttributesAll(sel, &attrs).Do(ctx); err != nil {
+		log.Printf("lastPhotoInDOM: document.quertSelectorAll:%s error %s", sel, err)
```
Comment (mpl) [inline main.go]: if you want to add that context, just add it to the returned error, as you've done below at L337? And let the caller be in charge of the logging.

File: main.go [hunk 4]
```diff
@@ -280,54 +278,70 @@ func (s *Session) firstNav(ctx context.Context) error {
...
+		return "", err
+	}
+	if len(attrs) == 0 {
+		return "", fmt.Errorf("lastPhotoInDOM: no elements match")
+	}
+
+	lastElement := attrs[len(attrs)-1]
+	href := lastElement["href"]
```
Comment (mpl) [inline main.go]: Is this guaranteed to exist, or can we end up with the empty string here?

File: main.go [hunk 5]
```diff
@@ -377,7 +391,8 @@ func doRun(filePath string) error {
 // navLeft navigates to the next item to the left
 func navLeft(ctx context.Context) error {
 	chromedp.KeyEvent(kb.ArrowLeft).Do(ctx)
-	chromedp.WaitReady("body", chromedp.ByQuery)
+	// Could wait for the location to change instead of this Sleep.
+	time.Sleep(200 * time.Millisecond)
```
Comment (mpl) [inline main.go]: yeah, I don't think we want arbitrary duration sleeps like that (as it's flaky, their effectiveness varying depending on the conditions), except when in a loop, when retrying things.

File: main.go (hunk 1 above)
Comment (daneroo) [reply main.go]: Yes, but notice that the `kb.PageDown,kb.End` are in the `for` loop, so they will happen as needed. 
The part that was removed (`kb.ArrowRight` and wait for `document.activeElement`) we're the parts removed because they no longer had any effect after scrolling.

File: main.go (hunk 2 above)
Comment (daneroo) [reply main.go]: Agreed. I will push the change.

File: main.go (hunk 3 above)
Comment (daneroo) [reply main.go]: Agreed!

File: main.go (hunk 5 above)
Comment (daneroo) [reply main.go]: I agree, I removed the `WaitReady("body",..)` because it had no effect, 

But I found that I needed a small delay, or my main loop would exit too soon, too often.

I did not want to tackle the *arbitrary* `Sleep(s)` in this PR, but I would like to propose an event based aproach. I have implemented this in a separate re-implmentation (node.js+puppeteer) which works really well.

 - **For navigation:** we can use `chromedp.ListenTarget` and get a callback when `page.EventNavigatedWithinDocument` or `target.EventTargetInfoChanged` occur, this is more reliable, and is much *tighter* from a timing perspective.
- **For download initiation:** Similarly, there is a `page.EventDownloadWillBegin` event which can obviate the directory watching loop in `(*Session) download()`

If you agree I can make a separate pull request to demonstrate this idea.

File: main.go (hunk 4 above)
Comment (daneroo) [reply main.go]: Yes, It is guarateed to exist, 
- first because `len(attrs)>0` (or we would have returned an error, (which is ignored for retry in the caller), 
- and the DOM Selector itself (`a[href^="./photo/"]`) separately guarantees that any returned `href` will start with "./photo/...", *(that is the meaning of the `^=` operator in the css selector)*

**Comment by daneroo** [comment]
I have pushed the two requested changes. (commit: *Prefer caller to log errors*)

Thanks for your patience, I am just getting used to working remotely, and in smaller increments!
I see the value.
I also have the `-headless` PR ready to go. As soon as we merge this, I'll rebase, test and submit

Also *Bonne année*!

File: main.go (hunk 5 above)
Comment (mpl) [reply main.go]: Sure, I'm very interested to see your approaches in another PR. But in the meantime, let's leave navLeft as it was. The fact that navLeft sometimes does not work for you is unrelated to the change that removes the screenshots, right?

**Comment by mpl** [comment]
> I have pushed the two requested changes. (commit: _Prefer caller to log errors_)
> 
> Thanks for your patience, I am just getting used to working remotely, and in smaller increments!
> I see the value.

yes, the smaller they are, the greater the chance I find a small slice on time to review one and that we make progress on them.
I was busy again this week-end, but I should have more time starting in a couple of days and thereafter.

> I also have the `-headless` PR ready to go. As soon as we merge this, I'll rebase, test and submit

excellent

> Finally: _Bonne année_!

Bonne année, bonne santé.

File: main.go (hunk 5 above)
Comment (mpl) [reply main.go]: So here's my empirical testing on this point:

1) If I remove your time.Sleep, I have the problem that it only gets one photo, and then it stops.
2) If re-add your time.Sleep, it indeed seems to help (but we don't want that).
3) If I re-add instead the previous code (i.e. chromedp.WaitReady("body", chromedp.ByQuery)), it also seems to work. So if what you say is true (that chromedp.WaitReady("body", chromedp.ByQuery) does nothing), maybe it still has the fortunate side-effect of slowing things down (and making navLeft work) just because we're spending some CPU cycles  doing "something"?

Comment (daneroo) [reply main.go]: Ok, I agree, that arbitray Sleep is *bad*, so until we have a *better condition* to wait for, I will restore navLeft as it was.

File: main.go [hunk 6]
```diff
@@ -280,54 +278,69 @@ func (s *Session) firstNav(ctx context.Context) error {
...
+		previousHref = lastHrefInDOM
 	}
 
-	if *verboseFlag {
-		log.Printf("Successfully jumped to the end")
+	// Now that we have stopped scrolling, select (focus) on the last element
+	// The element must be focused, so that navToLast can send "\n" to enter photo detail page
+	lastEltSel := fmt.Sprintf(`a[href="%s"]`, previousHref)
+	if err := chromedp.Focus(lastEltSel).Do(ctx); err != nil {
+		return err
```
Comment (mpl) [inline main.go]: I've just realized; doesn't this actually do the job of navToLast (except the actual navigation itself)? So couldn't we move that code to inside navToLast, so that it replaces there the loop that relies on the location change? And then we would only have to do one "\n" key event right after?

Alternatively, we could keep it here, and remove most of the the code in navToLast (it would just do the one "\n" event basically, but I don't like this solution as much as it does not separate as well each task.

Comment (daneroo) [reply main.go]: Yes, we could structure it that way.
but I also prefer keeping the tasks separate.

In my other experiment, I am trying to abstract the following:
- positioning: AlbumPage[First|Last],DetailPage[FirstLast]
- iterating: AlbumPage[Forward|Reverse], DetailPage[Forward|Reverse]
- itemCallback(item): [list,files,perkeep] 
  - *files and perkeep also have an optimistic variant which checks for existence before performing download*
- This way I can use the same outer loop for all combinations
- The iterators are using browser events [focuschanged,targetchanged] instead of polling and arbitrary `sleeps`. There is also an event fired for an initiated download, which obviates the need for the directory polling.

When I'm done finding the right abstractions I can port it back to Go/cdp.

Comment (mpl) [reply main.go]: > Yes, we could structure it that way.
> but I also prefer keeping the tasks separate.

Hmm, I'm not sure we understood each other, so just to be clear:
1) do you agree that most of what navToLast is doing is redundant with and already done by L309 and L310?
2) if yes, do you agree to move L309 and L310 to inside navToLast?

If I'm not making sense, I can add myself a commit on top of this PR so you can see what I mean?

> In my other experiment, I am trying to abstract the following:
> 
> * positioning: AlbumPage[First|Last],DetailPage[FirstLast]
> * iterating: AlbumPage[Forward|Reverse], DetailPage[Forward|Reverse]
> * itemCallback(item): [list,files,perkeep]
>   
>   * _files and perkeep also have an optimistic variant which checks for existence before performing download_
> * This way I can use the same outer loop for all combinations
> * The iterators are using browser events [focuschanged,targetchanged] instead of polling and arbitrary `sleeps`. There is also an event fired for an initiated download, which obviates the need for the directory polling.
> 
> When I'm done finding the right abstractions I can port it back to Go/cdp.

Comment (daneroo) [reply main.go]: Thanks for clarifying.
No, *I don't agree that most of what navToLast is doing is redundant*

We still need `navToLast` to work the way it does, because the selected element (309-310), is not guaranteed to be the *actual* last element, it is just *close* to the end.

File: main.go [hunk 7]
```diff
@@ -280,54 +278,69 @@ func (s *Session) firstNav(ctx context.Context) error {
...
 	}
 
+	if *verboseFlag {
+		log.Printf("Successfully jumped to the end: %s", previousHref)
```
Comment (mpl) [inline main.go]: Hmm, then I don't think we should have previousHref in this logging message, as it might lead one to believe that previousHref is the very last element of the page (which it might not be, as you've just explained).

File: main.go [hunk 8]
```diff
@@ -280,54 +278,69 @@ func (s *Session) firstNav(ctx context.Context) error {
...
+	var attrs []map[string]string
+	if err := chromedp.AttributesAll(sel, &attrs).Do(ctx); err != nil {
+		return "", err
+	}
+	if len(attrs) == 0 {
+		return "", fmt.Errorf("lastPhotoInDOM: no elements match")
+	}
+
+	lastElement := attrs[len(attrs)-1]
+	href := lastElement["href"]
```
Comment (mpl) [inline main.go]: href isn't used by anything, so you could save yourself one variable, and just

return lastElement["href"], nil

File: main.go (hunk 6 above)
Comment (mpl) [reply main.go]: Ok, understood, let's leave L309 and navToLast as they are for now.
Do you know why L309-L310 isn't enough to get the very last element though?

**Comment by daneroo** [comment]
> Do you know why L309-L310 isn't enough to get the very last element though?

Yes, it is because that css selector captures more that the elements that are visible on the screen. There is a complex virtual windowing/scrolling thing going on. It's a fancy DOM version of a buffer overrun!!! Reverse engineering is sooo kludgy.

I revisited all of my assumptions and I implemented a much simpler solution.
I think you are gonna like this.

- Removed the screen capture
- Remove the complex css selector stuff, (yuck, never liked that, too complicated 8-))
- Simply advance the selection with repeated `kb.RightArrow`,`kb.End`.
- Termination criteria is when active selection stops changing.

Pushing the update to the PR now.

**Comment by daneroo** [comment]
Ok, I think this is much better and simpler.

File: main.go [hunk 9]
```diff
@@ -280,49 +278,41 @@ func (s *Session) firstNav(ctx context.Context) error {
...
-	var previousScr, scr []byte
-	for {
-		chromedp.KeyEvent(kb.PageDown).Do(ctx)
-		chromedp.KeyEvent(kb.End).Do(ctx)
-		chromedp.CaptureScreenshot(&scr).Do(ctx)
-		if previousScr == nil {
-			previousScr = scr
-			continue
+		if *verboseFlag {
+			log.Printf("** navToEnd:activeElt %s %d", active, lastRepeated)
```
Comment (mpl) [inline main.go]: we don't really use that kind of telegraphic style, but rather full sentences.

log.Printf("Active element %s was seen %d times", active lastRepeated)

File: main.go [hunk 10]
```diff
@@ -280,49 +278,41 @@ func (s *Session) firstNav(ctx context.Context) error {
...
 		}
-		if bytes.Equal(previousScr, scr) {
+		if lastRepeated > 2 {
```
Comment (mpl) [inline main.go]: why is

    if lastRepeated > 1 {

not enough?

File: main.go [hunk 11]
```diff
@@ -280,49 +278,41 @@ func (s *Session) firstNav(ctx context.Context) error {
...
 			break
 		}
-		previousScr = scr
-		time.Sleep(tick)
+		prev = active
+		// time.Sleep(tick)
```
Comment (mpl) [inline main.go]: we don't keep commented out code.

File: main.go [hunk 12]
```diff
@@ -280,49 +278,41 @@ func (s *Session) firstNav(ctx context.Context) error {
...
 	}
-
 	if *verboseFlag {
-		log.Printf("Successfully jumped to the end")
+		log.Printf("Successfully jumped to the end: %s", active)
```
Comment (mpl) [inline main.go]: Same comment as last time, if active is not actually the very last element.

**Comment by mpl** [comment]
> > Do you know why L309-L310 isn't enough to get the very last element though?
> 
> Yes, it is because that css selector captures more that the elements that are visible on the screen. There is a complex virtual windowing/scrolling thing going on. It's a fancy DOM version of a buffer overrun!!! Reverse engineering is sooo kludgy.
> 
> I revisited all of my assumptions and I implemented a much simpler solution.
> I think you are gonna like this.
> 
> * Removed the screen capture
> * Remove the complex css selector stuff, (yuck, never liked that, too complicated 8-))

I don't remember if I told you, but before I did the screenshots method, one idea I had was to use the presence of their custom scrollbar on the right. As you've probably noticed, as long as you keep scrolling down, it stays visible. And if you stop, it disappears. But, when you reach the bottom of the page, even if you keep on scrolling, it eventually disappears. Unfortunately I did not find out how to "select" it.

> * Simply advance the selection with repeated `kb.RightArrow`,`kb.End`.
> * Termination criteria is when active selection stops changing.
> 
> Pushing the update to the PR now.

File: main.go [hunk 13]
```diff
@@ -280,49 +278,41 @@ func (s *Session) firstNav(ctx context.Context) error {
...
 		chromedp.KeyEvent(kb.ArrowRight).Do(ctx)
+		chromedp.KeyEvent(kb.End).Do(ctx)
 		time.Sleep(tick)
-		var ids []cdp.NodeID
-		if err := chromedp.Run(ctx,
-			chromedp.NodeIDs(`document.activeElement`, &ids, chromedp.ByJSPath)); err != nil {
-			return err
+
+		if err := chromedp.Evaluate(`document.activeElement.href`, &active).Do(ctx); err != nil {
+			time.Sleep(tick) // this extra sleep is important: after the kb.End, it sometimes takes a while for the active element to be reset
```
[the remark this answers is not in the document]
Comment (daneroo) [reply main.go]: Ok, for the comment, I will move to the line above. (I just pushed the update)

The reason the `continue` is preferable is that when the error condition occurs, (and enough time has passed, with the extra `Sleep`), the next thing that needs to happen is the `kb.ArrowRight`.

tl;dr

There are a lot of possible states. 
As far as I can tell (having experimented a lot last night):
- Initially, there is no active element, the ArrowRight will select the first photo. but the first `End` event does not actually perform the scroll until the ArrowRight has caused an element to be active (selected)
- On a subsequent iteration, the `End` event causes a scroll, and then there is *no* active element again (this is the tricky timing part)
  - if the scroll has stabilized (time-sensitive), then the next kb.ArrowRight will select the first photo at the new scroll position, (this allows advancement to continue from there.
  - however, if the scroll has not *stabilized*, the `ArrowRight` will have the effect of reselecting the first element back at the top of the page.
- Finally, after the `ArrowRight` has successfully selected an item at the new scroll position, all subsequent `kb.End` events have no effect.

So it turns that the way the loop is written is a pretty economical way of accomplishing everything we need. I did try to unroll every one of these steps, but it was less stable, and much more code.

File: main.go (hunk 9 above)
Comment (daneroo) [reply main.go]: Ok, fixed. I will keep that in mind.

File: main.go (hunk 10 above)
Comment (daneroo) [reply main.go]: It does happen, although very rarely, that the navigation (ArrowRight) stutters (does not advance). I discovered this after running the code hundreds of times in a burn-in test.

File: main.go (hunk 11 above)
Comment (daneroo) [reply main.go]: Yup, I missed that. Fixed it.

File: main.go (hunk 12 above)
Comment (daneroo) [reply main.go]: Now that the we are using `document.activeElement`, and `lastRepeated` in our termination criteria, we are actually sure that we do have the last element.

**Comment by daneroo** [comment]
Thanks for your attention to detail.

File: main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpl) [reply main.go]: > Ok, for the comment, I will move to the line above. (I just pushed the update)
> 
> The reason the `continue` is preferable is that when the error condition occurs, (and enough time has passed, with the extra `Sleep`), the next thing that needs to happen is the `kb.ArrowRight`.
> 
> tl;dr
> 
> There are a lot of possible states.
> As far as I can tell (having experimented a lot last night):
> 
> * Initially, there is no active element, the ArrowRight will select the first photo. but the first `End` event does not actually perform the scroll until the ArrowRight has caused an element to be active (selected)
> * On a subsequent iteration, the `End` event causes a scroll, and then there is _no_ active element again (this is the tricky timing part)
>   
>   * if the scroll has stabilized (time-sensitive), then the next kb.ArrowRight will select the first photo at the new scroll position, (this allows advancement to continue from there.
>   * however, if the scroll has not _stabilized_, the `ArrowRight` will have the effect of reselecting the first element back at the top of the page.

I don't understand. Isn't it pretty bad to let that happen? Does this case still happen with the current patch?

> * Finally, after the `ArrowRight` has successfully selected an item at the new scroll position, all subsequent `kb.End` events have no effect.
> 
> So it turns that the way the loop is written is a pretty economical way of accomplishing everything we need. I did try to unroll every one of these steps, but it was less stable, and much more code.

File: main.go (hunk 12 above)
Comment (mpl) [reply main.go]: well, then my other remark from the previous patch stands again I think ;P
i.e.: are we now not redundant with some of what navToLast does?

Comment (daneroo) [reply main.go]: You are right that It is redundant with some of what `navToLast` does, but certainly compatible with it.

I thought we were trying to keep the change minimal (the PR is trying to replace the screenshot stuff).

Do you want to also modify `navToLast` to only what is essential to work with this new version of `navToEnd`?

Comment (mpl) [reply main.go]: > You are right that It is redundant with some of what `navToLast` does, but certainly compatible with it.
> 
> I thought we were trying to keep the change minimal (the PR is trying to replace the screenshot stuff).

Yes, ideally we always want to keep a PR about one change and one change only, and not about several unrelated changes. However, if your fix to remove the screenshots makes navToLast obsolete and useless (which I think it does), then it's actually a good thing to remove it as part of the same PR.

> Do you want to also modify `navToLast` to only what is essential to work with this new version of `navToEnd`?

Yep, or even better, let's fully remove it if we can. But if you're tired of this review, I can do it in a subsequent PR, no worries.

File: main.go (hunk 10 above)
Comment (mpl) [reply main.go]: hmm, then can you add a comment about it please?

File: main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mpl) [reply main.go]: I believe you tested that part and it works, but I'm still confused about why and how it works, so I'll have to run a few experiments myself with it when I have a bit of time.

[the remark this answers is not in the document]
Comment (mpl) [reply main.go]: FYI, I'm still experimenting on that part, because I don't like that we still rely on an arbitrary timeout.

**Comment by daneroo** [comment]
I just update the PR:
- integrate navToLast into navToEnd
- added waiting logic to navLeft (because it fails too often on first image especially)

File: main.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (daneroo) [reply main.go]: Ok, looking forward to your results.
I just pushed an update (See below)

[the remark this answers is not in the document]
Comment (mpl) [reply main.go]: I found an algorithm that I think is simpler to understand (because it uses left arrow instead of right arrow), but it still suffers from needing one sleep, which sucks.

basically, what I want is:

```
	var first string
	for {
		chromedp.KeyEvent(kb.ArrowLeft).Do(ctx)
		if err := chromedp.Evaluate(`document.activeElement.href`, &first).Do(ctx); err != nil {
			continue
		}
		if first == "" {
			continue
		}
		break
	}

	var current string
	for {
		chromedp.KeyEvent(kb.End).Do(ctx)
		// TODO: get rid of that sleep
		time.Sleep(tick)
		for {
			chromedp.KeyEvent(kb.ArrowLeft).Do(ctx)
			if err := chromedp.Evaluate(`document.activeElement.href`, &current).Do(ctx); err != nil {
				continue
			}
			break
		}
		if current == first {
			continue
		}
		break
	}
```
And yes, I know that does not get us to the very last element. but I don't really care, that's easy to fix. Or we can just keep the call to navToLast after, whatever.

Isn't there anything that we can do after a call to 
    chromedp.KeyEvent(kb.End).Do(ctx)
that would allow us to actively detect and wait for the jump to be done?

File: main.go [hunk 14]
```diff
@@ -370,9 +344,35 @@ func doRun(filePath string) error {
...
+
+		if err := chromedp.Location(&location).Do(ctx); err != nil {
+			return err
+		}
+		if location != prevLocation {
+			log.Printf("navLeft break at it:%d", i)
+			break
+		}
+		time.Sleep(miniTick)
+	}
```
Comment (mpl) [inline main.go]: let's worry about that in another PR please.

Comment (daneroo) [reply main.go]: Ok, I will remove that change.

Unfortunately, the original navLeft does not work for me, wich makes testing rather hard.

I think it's time we just close this PR.
I am pretty lost in the threaded comments.

I'd still love to help out somehow, but I am not sure what the best way to have a productive conversation is.

Comment (daneroo) [reply main.go]: Ok, I will remove that change.

Unfortunately, the original navLeft does not work for me, wich makes testing rather hard.

I think it's time we just close this PR.
I am pretty lost in the threaded comments.

I'd still love to help out somehow, but I am not sure what the best way to have a productive conversation is.

Comment (mpl) [reply main.go]: > Ok, I reverted the change to `navLeft`.
> Unfortunately, the original `navLeft` does not work for me, which makes testing rather hard. _(it falsely terminates early too often)_.

Can you create another PR for that fix please? we don't have to wait for the current one to be over. It can all be done in parallel.

> I'm also a bit lost in the threaded comments. Perhaps I'm not using the review tooling correctly.

I suspect it's both. We probably got carried away too many times, and github can be weird at times (I'm still much more comfortable myself in gerrit tbh). I suggest reloading the page, taking a big breath and rereading it all from the start, noting as you go along what is actually still a conversation, and what is not. :-)

> If you feel this is dragging on, we can also just close the PR.

Let's not close it please. I am still hopeful that we can make one (variant) of the solutions you provided work, i.e. replacing the screenshots with another one that is less gross, but that also does not use arbitrary sleeps. But if you are bored with that problem, no worries, I'll just keep on iterating myself on that PR. And of course feel free to open other (independent) PRs in parallel.

Comment (daneroo) [reply main.go]: Cool. I did a separate PR for `-headless` flag.
I will look at navLeft next, and I can circle back to this navToLast/End again after.
